### ISSUE TITLE
Tax calculation wrong after applying coupon code (Major Bug)

### ISSUE THREAD

**Issue by sudheeshks1985** [magento/magento2#8964]
### Preconditions

1. Magento 2.1.5

### Steps to reproduce

1. Stores->Configuration-> Sales->Tax : Set all values including tax
2. Stores->Tax Rules : Set VAT of 16%
3. Marketing-> Cart Price Rule : Set 10% offer and apply to shipping amount also

### Expected result

1. Cart Check out page Shipping calculator : After applying 10% coupon code tax calculation is getting wrong. Here subtotal is K 1160 and K 10 is shipping charge.So 10% offer will be K 117 and Order Total Inclusive of Tax is K 1053. 16% VAT of K 1053 should be K 145.24. Also Order Total Exclusive of Tax should be K 907.76

### Actual result

VAT is 145.38 and Order Total Exclusive of Tax is 907.62.

1. [Screenshot, logs]
![shipping_with_giftcoupon](https://cloud.githubusercontent.com/assets/4659674/24167408/ae63d6f6-0e9c-11e7-9c61-b75e5e3f830a.png)

**Comment by sudheeshks1985** [comment]
I am facing same issue in magento 2.1.5 basic installation also.

16% VAT is applied for products and shipping charge.

Apply to Shipping Amount is YES for Coupon code OFFER10

Please check the attached image

Please check this issue ASAP. I cant able to go to live due to this major issue.
![discount_coupon_magento215](https://cloud.githubusercontent.com/assets/4659674/24233641/8e6b8894-0fb9-11e7-92dc-d48e599ae11b.png)

**Comment by sudheeshks1985** [comment]
Nobody faced this issue before..?? Please check and fix this issue asap.. Or guide me to do a fix

**Comment by Pigau** [comment]
I have the same issue on Magento 2.1.5 : 
If I have : 
subtotal : **10,00 $ including tax**
shipping: **5 $ including tax**
tax : **20 %**
discount: **15 $**

I still have to pay 0.85 $ (tax amount of my shipping fees)

**Comment by thomasnordkvist** [comment]
I have the same issue on Magento 2.1.6

**Comment by tjack88** [comment]
Has a fix been pushed for this yet? Can you help @ishakhsuvarov , it is really quite urgent and pressing that this is sorted

@thomasnordkvist, @Pigau, @sudheeshks, any of you guys found a solution?, or have found a temporary workaround/hack?

**Comment by ishakhsuvarov** [comment]
@gold2040 I am not aware of any solution implemented by now. Did you check if issue exists in the `develop` branch?
Please upvote the issue, it may speed up the process.
Thank you.

**Comment by JacobDrummond** [comment]
Related to #10067 ?

**Comment by magento-engcom-team** [comment]
@sudheeshks, thank you for your report.
We've created internal ticket(s) MAGETWO-75553 to track progress on the issue.

**Comment by TomMeijnaerts** [comment]
ill try to give it a go

**Comment by TomMeijnaerts** [comment]
the issue is related with the setting at **Stores > Configuration > Sales > Tax**
<img width="650" alt="screen shot 2017-09-26 at 16 11 25" src="https://user-images.githubusercontent.com/7273436/30864953-7e7f56d6-a2d5-11e7-96aa-2140b76dcfad.png">

When do filling in the Tax class for shipping the calculations are correct 

- i verified with the same tax class as the product
- i verified with a different tax class as the product (0% and another value)

```
**Calculations with shipping 0%**
Product - excl = 1000
Product - Incl = 1160

Shipping - excl = 5
Shipping - incl = 5

Discount - product = 116
Discount - shipping = 0,5
Discount - Total Excl = 100,5 (10% of the Product-excl + 10% Shipping-excl)
Discount - Total Incl = 116,5

Total Excl = 1000 + 5 - 100,5 = 904,5
Total Incl = 1160 + 5 - 116,5 = 1048,5
Total Tax = Total Incl - Total Excl = 144
```

```
**Calculations with shipping 16%**
Product - excl = 1000
Product - Incl = 1160

Shipping - excl = 4,31
Shipping - incl = 5

Discount - product = 116
Discount - shipping = 0,5
Discount - Total Excl = 100,43 (10% of the Product-excl + 10% Shipping-excl)
Discount - Total Incl = 116,5

Total Excl = 1000+4,31-100,43 = 903,88
Total Incl = 1160 + 5 - 116,5 = 1048,5
Total Tax = Total Incl - Total Excl = 144,62
```

in both situations it calculates the TAX correctly (given into account that the discount excl amount has 2 tax classes, 1 for the product and a different one for the shipping part)

When not setting the Tax class for shipping (default option), the function getShippingTaxClass in Magento\Tax\Model\Config will return 0. Because 0 is not a valid tax-class the calculation is doing something that it is not supposed to. 

Once you set a value for the shipping tax and you want to reproduce the original problem, the default option follows the last know configuration so it could also be a default db setting problem. (WIP)

**Comment by ghost** [comment]
Has this been Fixed yet in Latest magento verson ?

**Comment by AlexanderHuyghebaert** [comment]
Is there already a solution for this? Fix/Patch...

**Comment by mpalasis** [comment]
Discount w/ and w/o tax has been buggy since Magento 1 (and it is still Buggy as of Mage 1.9.3.8).
I really had trouble sorting this out a few years back.
This and the no-duplicate issue is blocking me from upgrading to Mage 2.x
Any update on this yet, Please!?

**Comment by axovel** [comment]
This is quite annoying that how such a critical issue can be opened for 18th months since its reported. Many people are facing issue with Tax calculation and loosing their customer.  The same problem is with shopping cart price rules also.

**Comment by Smilefounder** [comment]
Hello Magento team, any fix patch that able to apply on 2.2.5? 

Best regards.

**Comment by Smilefounder** [comment]
Able to reproduce the issue with 2.2.6.

**Comment by bawwwa** [comment]
Hi, 

I have tried this and it seems to be working. 

Magento\SalesRule\Model\Validator::processShippingAmount(Address $address)

from
`
        $shippingAmount = $address->getShippingAmountForDiscount();
        if ($shippingAmount !== null) {
            $baseShippingAmount = $address->getBaseShippingAmountForDiscount();
        } else {
            $shippingAmount = $address->getShippingAmount();
            $baseShippingAmount = $address->getBaseShippingAmount();
        }
`
to
`
        $shippingAmount = $address->getShippingAmountForDiscount();
        if ($shippingAmount !== null) {
            $baseShippingAmount = $address->getBaseShippingAmountForDiscount();
        } else {
            $shippingAmount = $address->getShippingInclTax();
            $baseShippingAmount = $address->getBaseShippingInclTax();
        }
`

Thanks!

**Comment by marcusirgens** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @marcusirgens. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @marcusirgens, here is your [Magento instance](https://i-8964-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-8964-2-3-develop.instances.magento-community.engineering/admin](https://i-8964-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by marcusirgens** [comment]
Confirming this issue is still valid on 2.3-develop.

*Expected result:*
A discount that also applies to shipping costs will correctly apply:
For a flat rate shipping with a cost of 100 including 25% VAT (20), the shipping price when applying a 50% 
 cart rule should be 50, with 10 being the tax amount
*Actual result*
The discount is 60, with 20 being the tax amount.

**Comment by azambon** [comment]
The change proposed by @bawwwa is close to the solution that I found.

The issue seems to be that in the Magento\Tax\Model\Sales\Total\Quote\CommonTaxCollector::processShippingTaxInfo() method, Magento sets the "shipping_amount_for_discount" and "base_shipping_amount_for_discount" values on the total object, whereas in Magento\SalesRule\Model\Validator::processShippingAmount() Magento tries to read those values from the address.

Since the processShippingAmount() method does not receive the total object, I fixed the issue in the caller method, Magento\SalesRule\Model\Quote\Address\Total\ShippingDiscount::collect() by manually copying the values from the total to the address.
I added the lines:

```
$address->setShippingAmountForDiscount($total->getShippingAmountForDiscount());
$address->setBaseShippingAmountForDiscount($total->getBaseShippingAmountForDiscount());
```

just before the line

`$this->calculator->processShippingAmount($address);`

**Comment by bwat-dev** [comment]
**Same issue on 2.3.2 :(**
The discount calculation on shipping cost is wrong whit a percent discount.

Ex : 
Cart 10€ inluding tax
Delivery 10€ including tax
Tax 20% on products and shipping
Discount 20%
Settings : 
- apply tax : before discount
- apply discount on price :  tax included

**Expected result :**
Price after discount : 16€
= 20  - 10x0,2 - 10x0,2 
[Total] - [Price tax included] x 20%  - [Shipping tax included] x 20%

**Actual result :**
Price after discount : 16,40€
= 20  - 10x0,2 - 8x0,2 
[Total] - [Price tax included] x 20%  - **[Shipping without tax]** x 20%

**Comment by drazulay** [comment]
Anybody home?

**Comment by gitmoorka** [comment]
any news on when the fix would be available for ordinary users?

**Comment by engcom-Hotel** [comment]
Hello @sudheeshks,
Trying to reproduce the issue.
Magento version: 2.4-develop
PHP version: 7.4
ElasticSearch version: 7
MySQL version: 8.0

1. "Stores->Configuration-> Sales->Tax : Set all values including tax" - passed, not clear step. Default config used. 
2. Stores->Tax Rules->Create a new Rule with VAT of 10% (10 instead of 16 easier to count)
3. Marketing-> Cart Price Rule->Create new cart rule with 10% discount with coupon code (Apply to Shipping Amount => Yes)
4. Add products to the cart and apply coupon code.
Before discount
![8964_before_coupon](https://user-images.githubusercontent.com/51681618/103407427-54b82100-4b67-11eb-9748-2ca62ac58a0f.png)
After discount
![8964_after_coupon](https://user-images.githubusercontent.com/51681618/103407442-5c77c580-4b67-11eb-9ec1-01a8c7e4d8b6.png)
Calculation 10% tax on 100 = 10, 10% tax on 90 (after discount) = 9.
Looks correct. Could you please let me know if you are able to reproduce it on 2.4-develop and if yes, please provide detailed steps to do it?
Thank you.

**Comment by azambon** [comment]
@engcom-Hotel looks like a fix very similar to the one that I proposed in an earlier comment was applied about one year ago: https://github.com/magento/magento2/commit/a7928e2ec12f685159e6e92799d7831c4f0014be#diff-db8fbaf4cd61c34bed585db10bff06f1ac63442df3422c5d672d230fb7c48e49

It was fixed in 2.3.5, as far as I can tell
https://github.com/magento/magento2/blob/2.3.5/app/code/Magento/SalesRule/Model/Quote/Address/Total/ShippingDiscount.php

**Comment by engcom-Hotel** [comment]
@azambon, thanks for the info.
Closing the issue, please feel free to reopen it if you still have this issue.
